Add support for custom rosdep URLs in devel builds (fixes #87)

jjekircp opened the pull request:
I went by [tfoote's comment on issue #87](https://github.com/ros-infrastructure/ros_buildfarm/issues/87#issuecomment-117556312) suggesting to use the _config section of the targets in the release build file to get a list of custom rosdep URLs.
I then pass that variable through to the devel job, where I wget those URLs into /etc/ros/rosdep/sources.list.d. If no custom rosdep URLs are provided the job runs "rosdep init" as usual.

Here is the targets: entry I use in ros_buildfarm_config/indigo/source_build.yml that makes this work:
```
targets:
  _config:
    custom_rosdep_urls:
      - http://my-git-server/rosdistro-mine/raw/master/rosdep/sources.list.d/20-default.list
```

gregwjacobs commented:
+1

tfoote commented:
Thanks this looks reasonable. I want to get a moment to test it before I merge it. 

@ddimarco @dejanpan FYI.

dirk-thomas commented:
The PR should also update the documentation and describe the new configuration options: https://github.com/ros-infrastructure/ros_buildfarm/blob/master/doc/configuration_options.rst

dirk-thomas commented on ros_buildfarm/common.py:
```diff
@@ -51,6 +53,8 @@ def get_repositories_and_script_generating_key_files(
             len(build_file.repository_keys)
         repository_urls += build_file.repository_urls
         repository_keys += build_file.repository_keys
+        if hasattr(build_file, "custom_rosdep_urls"):
```
Please use single quotes wherever possible. Same multiple times below.

Regarding the `hasattr` I will comment below.

dirk-thomas commented on ros_buildfarm/config/source_build_file.py:
```diff
@@ -92,6 +92,10 @@ def __init__(self, name, data):
 
+        if '_config' in data['targets']:
+            if "custom_rosdep_urls" in data['targets']['_config']:
+                self.custom_rosdep_urls = data['targets']['_config']['custom_rosdep_urls']
```
Please add an assertion to ensure the expected type.

dirk-thomas commented on ros_buildfarm/templates/devel/devel_create_tasks.Dockerfile.em:
```diff
@@ -20,6 +20,7 @@ RUN useradd -u @uid -m buildfarm
     add_source=False,
 ))@
 
+
```
Please avoid white space only changes.

dirk-thomas commented on ros_buildfarm/templates/devel/devel_create_tasks.Dockerfile.em:
```diff
@@ -28,12 +29,16 @@ RUN useradd -u @uid -m buildfarm
-RUN rosdep init
+
+@(TEMPLATE(
+    'snippet/add_custom_rosdeps.Dockerfile.em',
```
Since the task of the snippet is more general then adding custom rosdeps I would propose to call it "rosdep_init.Dockerfile.em".

dirk-thomas commented on scripts/devel/run_devel_job.py:
```diff
@@ -9,6 +9,7 @@
 from ros_buildfarm.argument import \
     add_argument_distribution_repository_key_files
 from ros_buildfarm.argument import add_argument_distribution_repository_urls
+from ros_buildfarm.argument import add_argument_custom_rosdep_urls
```
Please insert the import in alphabetical order following PEP 8.

jjekircp commented:
Thanks for the prompt feedback. I've pushed an update addressing all of @dirk-thomas's comments.

dirk-thomas commented on ros_buildfarm/config/source_build_file.py:
```diff
@@ -92,6 +92,11 @@ def __init__(self, name, data):
 
+        if '_config' in data['targets']:
+            if 'custom_rosdep_urls' in data['targets']['_config']:
+                self.custom_rosdep_urls = data['targets']['_config']['custom_rosdep_urls']
```
Please always add the member variable to this class. E.g. with a default value of an empty list.

Then you also don't have to use `hasattr` above but can unconditionally access the member.

dirk-thomas commented on doc/configuration_options.rst:
```diff
@@ -81,7 +81,8 @@ A set of options which can be used in any build file.
   The second level key contains the OS code name.
   The third level key contains the CPU architecture.
-  The third level value is always empty.
+  The third level value is always empty. The first level key "_config" is
```
Please start each sentence on a new line (see https://github.com/ros2/ros2/wiki/Developer-Guide#markdown for the rational).

Same below.

jjekircp commented:
Updated again.
This wasn't called out specifically, but after taking a look at the Markdown Developer Guide linked above, I marked the yaml code block in the docs as yaml.
I added the custom_rosdep_urls variable to the BuildFile class, not the SourceBuildFile class. I hope this was the right choice; it ensures that the ``get_repositories_and_script_generating_key_files()`` function in common.py can access the member in all cases, including when processing a doc or release build file.

dirk-thomas commented on doc/configuration_options.rst:
```diff
@@ -235,6 +237,13 @@ The following options are valid in version ``2`` (beside the generic options):
+* ``custom_rosdep_urls``: a list of URLs containing rosdep sources.list.d entry
+ files that are downloaded into /etc/ros/rosdep/sources.list.d at the beginning
+  of the devel job.
+  If unset, the job simply runs _rosdep init_ instead.
```
The indentation is inconsistent in this bullet which breaks the rendering.

dirk-thomas commented on doc/custom_rosdistro.rst:
````diff
@@ -124,6 +124,17 @@ rosdep database:
+  _config:
+    custom_rosdep_urls:
+      - https://raw.githubusercontent.com/YOUR_FORK/rosdistro/master/rosdep/sources.list.d/20-default.list
+```
````
This block seems to not render as you might expect it to.

jjekircp replied:
Think I have it fixed now. I didn't realize at first that the previewer I was using locally (based on pandoc) doesn't treat reStructuredText in exactly the same way as GitHub does.

dirk-thomas commented on ros_buildfarm/config/build_file.py:
```diff
@@ -37,6 +37,8 @@ class BuildFile(object):
         self.name = name
 
+        self.custom_rosdep_urls = []
+
```
Please move this to the `SourceBuildFile` (above line 95) for now. Once any other build file supports additional configuration we can generalize it then.

jjekircp replied:
Can do, but will then need to change ``get_repositories_and_script_generating_key_files()`` in common.py as some build_file objects passed into it will have ``custom_rosdep_urls`` defined and some will not (because that function gets called on release, source, and doc build files). Is ``hasattr`` the best way of ensuring I don't try to access an attribute that isn't defined or is there a better way that I should use instead?

dirk-thomas replied:
Sounds good to me to use `hasattr` in this case.

dirk-thomas commented:
LGTM but I have not tried to run / use it. Waiting for @tfoote to get to test it as mentioned above (https://github.com/ros-infrastructure/ros_buildfarm/pull/100#issuecomment-141563839).

ddimarco commented on ros_buildfarm/config/source_build_file.py:
```diff
@@ -92,6 +92,12 @@ def __init__(self, name, data):
+        if '_config' in data['targets']:
+            if 'custom_rosdep_urls' in data['targets']['_config']:
+                self.custom_rosdep_urls = data['targets']['_config']['custom_rosdep_urls']
+                assert.isinstance(self.custom_rosdep_urls, list)
```
I think this should read `assert isinstance(...)`, i.e. without the dot.

ddimarco commented on ros_buildfarm/templates/snippet/rosdep_init.Dockerfile.em:
```diff
@@ -0,0 +1,8 @@
+RUN (cd /etc/ros/rosdep/sources.list.d/ && wget --progress=dot @url )
+@[end for]@
+@[if not custom_rosdep_urls]@
+RUN rosdep init
```
I think we need `rosdep init` here regardless of whether `custom_rosdep_urls` is set.

jjekircp replied:
What does `rosdep init` do other than create the /etc/ros/rosdep/sources.list.d/20-default file from the default one on ros/rosdistro? It didn't seem to me like it did anything else, which is why I did it that way.
It'd be easy enough to change it to run `rosdep init` first and then overwrite 20-default.list with the user-specified one if desired though.

tfoote replied:
I would suggest running rosdep init either way. And not override the default sources, but add another file side by side with the additional sources. You should be able to set precidence using the number at the front of the file.

jjekircp replied:
Ok, I'll update that section to unconditionally run rosdep init.
I would prefer to keep the ability to override the default sources. Specifically I would like to fetch all of my rosdep sources from my forked rosdistro repo (whether or not they are identical to the public one). The primary motivation for this is that my rosdistro repo is hosted on an internal git server rather than GitHub, and so it's insulated from issues with GitHub or Internet connectivity at the site.

I can update the docs so it's clear to users that they can either use a combination of the default sources and their own custom sources file, or avoid the default sources entirely and link to the file in their forked rosdistro repo.

ddimarco commented:
Apart from the comment above, it works for me.

mistoll commented:
Tested this and it works for me.

tfoote commented:
@ddimarco @mistoll thanks for testing
